lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp

```diff
@@ -26,123 +28,144 @@ using namespace swift;
+}
 
-    // Then remove any begin/end borrow that we found. These are unneeded since
-    // the lifetime guarantee from the argument exists above and beyond said
-    // scope.
-    while (!NewBorrowInsts.empty()) {
-      SILInstruction *I = NewBorrowInsts.pop_back_val();
-      if (auto *BBI = dyn_cast<BeginBorrowInst>(I)) {
-        // Any copy_value that is used by the begin borrow is added to the
-        // worklist.
-        for (auto *BBIUse : BBI->getUses()) {
-          if (auto *BBIUseCopyValue =
-                  dyn_cast<CopyValueInst>(BBIUse->getUser())) {
-            Worklist.push_back(BBIUseCopyValue);
-          }
-        }
+static bool performGuaranteedCopyValueOptimization(CopyValueInst *cvi) {
+  SmallVector<SILValue, 16> borrowIntroducers;
 
-        // First go through and eliminate all end borrows.
-        SmallVector<EndBorrowInst *, 4> endBorrows;
-        copy(BBI->getEndBorrows(), std::back_inserter(endBorrows));
-        while (!endBorrows.empty()) {
-          endBorrows.pop_back_val()->eraseFromParent();
-          ++NumEliminatedInsts;
-        }
+  // Whitelist the operands that we know how to support and make sure
+  // our operand is actually guaranteed.
+  if (!canHandleOperand(cvi->getOperand(), borrowIntroducers))
+    return false;
 
-        // Then eliminate BBI itself.
-        BBI->replaceAllUsesWith(BBI->getOperand());
-        BBI->eraseFromParent();
-        ++NumEliminatedInsts;
+  // Then go over all of our uses. Find our destroying instructions
+  // and make sure all of them are destroy_value. For our
+  // non-destroying instructions, make sure that they accept a
+  // guaranteed value. After that, make sure that our destroys are
+  // within the lifetime of our borrowed values.
+  SmallVector<DestroyValueInst *, 16> destroys;
+  for (auto *op : cvi->getUses()) {
+    // We know that a copy_value produces an @owned value. Look
+    // through all of our uses and classify them as either
+    // invalidating or not invalidating. Make sure that all of the
+    // invalidating ones are destroy_value since otherwise the
+    // live_range is not complete.
+    auto map = op->getOwnershipKindMap();
+    auto constraint = map.getLifetimeConstraint(ValueOwnershipKind::Owned);
+    switch (constraint) {
+    case UseLifetimeConstraint::MustBeInvalidated:
+      // And we have a destroy_value, track it and continue.
+      if (auto *dvi = dyn_cast<DestroyValueInst>(op->getUser())) {
+        destroys.push_back(dvi);
         continue;
       }
+      // Otherwise, we found a non-destroy value invalidating owned
+      // user... This is not an unnecessary live range.
+      return false;
+    case UseLifetimeConstraint::MustBeLive:
+      // Ok, this constraint can take something owned as live. Lets
+      // see if it can also take something that is guaranteed. If it
+      // can not, then we bail.
+      if (!map.canAcceptKind(ValueOwnershipKind::Guaranteed)) {
```

Comment:
[design] I really can't mentally map the code that's using these physical maps to concepts that I normally reason about. i.e. If I wanted to know how this would work on a concrete SIL example, I would have to construct a test case by hand and run it through the code. (that's the problem with too much indirection). I *think* this is what the code is trying to do (eliding some details), but not sure:
```
switch (operand.ownershipEffect()):
case Consuming:
  if (isa<Destroy>(operand.getUser()))
    continue;
  return false;

case NonConsuming:
  // A point-in-time use can always handle +1 or +0.
  continue;

case Forwarding:
  // This instruction takes ownership of the copy. We don't currently
  // recurse through them.
  return false;
}
```
If we wanted to constrain some non-consuming users to *only* handle owned
values, then we could introduce a separate "use constraint" API for
that. But I suspect we don't want to allow that anyway.